File: src/main/java/com/shaposhnyk/graphql/adfacade/IEntityFactory.java
package com.shaposhnyk.graphql.adfacade;

import com.google.common.base.Function;
import graphql.schema.DataFetchingEnvironment;
import graphql.schema.GraphQLFieldDefinition;
import graphql.schema.GraphQLType;

import javax.annotation.Nonnull;

public interface IEntityFactory<R> {
  /** GraphQL Object definition */
  @Nonnull
  GraphQLType getObjectDefinition();

  /**
   * Fetches values according to the parameters specified in env
   *
   * @return iterator from fetched values
   */
  @Nonnull
  Iterable<R> fetch(@Nonnull DataFetchingEnvironment env);

  /**
   * @param fieldName
   * @param fetcher
   * @return builder for a list of objects from factory
   */
  @Nonnull
  GraphQLFieldDefinition.Builder newListFieldDefinition(
      @Nonnull String fieldName, @Nonnull Function<DataFetchingEnvironment, Iterable<R>> fetcher);

  @Nonnull
  default GraphQLFieldDefinition.Builder newListFieldDefinition(@Nonnull String fieldName) {
    return newListFieldDefinition(fieldName, this::fetch);
  }
}

File: src/main/kotlin/com/shaposhnyk/graphql/adfacade/IEntityFactory.kt
package com.shaposhnyk.graphql.adfacade

import graphql.schema.DataFetchingEnvironment
import graphql.schema.GraphQLFieldDefinition
import graphql.schema.GraphQLType

interface IEntityFactory {
    /**
     * GraphQL Object definition
     */
    fun objectDefinition(): GraphQLType

    fun listFieldDefinition(fieldName: String, fetcher: (DataFetchingEnvironment) -> Any): GraphQLFieldDefinition.Builder

    /**
     * Fetches a list of instances of given entity
     */
    fun fetchList(env: DataFetchingEnvironment): Iterable<Any>

    fun listFieldDefinition(fieldName: String) = listFieldDefinition(fieldName, this::fetchList)
}